cicecore/cicedynB/dynamics/ice_dyn_shared.F90

```diff
@@ -1358,144 +1905,1025 @@ subroutine strain_rates (nx_block,   ny_block,   &
+         DeltaT       ! strain rates at the T point
+
+      ! local variables
+
+      real (kind=dbl_kind) :: &
+         ldivT     , &
+         ltensionT , &
+         lshearT       ! local values
+
+      integer (kind=int_kind) :: &
+         ij, i, j                  ! indices
+         
+      character(len=*), parameter :: subname = '(strain_rates_T)'
+
+      !-----------------------------------------------------------------
+      ! strain rates
+      ! NOTE these are actually strain rates * area  (m^2/s)
+      !-----------------------------------------------------------------
+
+      if (present(divT)    ) divT    (:,:) = c0
+      if (present(tensionT)) tensionT(:,:) = c0
+      if (present(shearT)  ) shearT  (:,:) = c0
+      if (present(deltaT)  ) deltaT  (:,:) = c0
+
+      do ij = 1, icellt
+         i = indxti(ij)
+         j = indxtj(ij)
+
+         ! divergence  =  e_11 + e_22
+         if (present(deltaT) .or. present(divT)) then
+            ldivT    = dyE(i,j)*uvelE(i  ,j  ) - dyE(i-1,j)*uvelE(i-1,j  ) &
+                     + dxN(i,j)*vvelN(i  ,j  ) - dxN(i,j-1)*vvelN(i  ,j-1)
+            if (present(divT)) then
+               divT(i,j) = ldivT
+            endif
+         endif
+
+         ! tension strain rate  =  e_11 - e_22
+         if (present(deltaT) .or. present(tensionT)) then
+            ltensionT= (dyT(i,j)**2)*(uvelE(i,j)/dyE(i,j) - uvelE(i-1,j)/dyE(i-1,j)) &
+                     - (dxT(i,j)**2)*(vvelN(i,j)/dxN(i,j) - vvelN(i,j-1)/dxN(i,j-1))
+            if (present(tensionT)) then
+               tensionT(i,j) = ltensionT
+            endif
+         endif
+
+         ! shearing strain rate  =  2*e_12
+         if (present(deltaT) .or. present(shearT)) then
+            lshearT  = (dxT(i,j)**2)*(uvelN(i,j)/dxN(i,j) - uvelN(i,j-1)/dxN(i,j-1)) &
+                     + (dyT(i,j)**2)*(vvelE(i,j)/dyE(i,j) - vvelE(i-1,j)/dyE(i-1,j))
+            if (present(shearT)) then
+               shearT(i,j) = lshearT
+            endif
+         endif
       
-      subroutine viscous_coeffs_and_rep_pressure (strength,  tinyarea, &
-                                                  Deltane,   Deltanw,  &
-                                                  Deltasw,   Deltase,  &
-                                                  zetax2ne,  zetax2nw, &
-                                                  zetax2sw,  zetax2se, &
-                                                  etax2ne,   etax2nw,  &
-                                                  etax2sw,   etax2se,  &
-                                                  rep_prsne, rep_prsnw,&
-                                                  rep_prssw, rep_prsse,&
-                                                  capping)
+         ! Delta (in the denominator of zeta, eta)
+         if (present(deltaT)) then
+            DeltaT  (i,j) = sqrt(ldivT**2 + e_factor*(ltensionT**2 + lshearT**2))
+         endif
+
+      enddo
+
+      end subroutine strain_rates_T
```

Comment:
Again, if we do inline strain_rates_U and _T, we lose code reuse.   We are going to have to make some decisions about code reuse and readability.  I think we should assume this is not a issue of performance until we have some proof.

Reply:
This is not so much about readability as about computational efficiency.  These calls are in the subcycling loop, where if statements can matter.  Improving readability is a plus.  Here are some suggestions, which might or might not work.

T:
stressC_T calls strain_rates_T to get divT, tensionT.
stressCD_T calls strain_rates_T to get divT, tensionT, shearT, DeltaT.
deformations_T calls strain_rates_T to get divT, tensionT, shearT, DeltaT.

Since divT and tensionT are calculated every time, those if statements can be removed.  A wrapper routine that gets all 4 could call an inner routine that just does divT and tensionT, which could also be called directly when shear and Delta are not needed.

U:
evp (grid_ice=C) calls strain_rates_U to get shearU, apparently separate because it needs a halo update.
stressC_U calls strain_rates_U to get DeltaU.
stressCD_U calls strain_rates_U to get divU, tensionU, shearU, DeltaU.

strain_rates_U is a bit more complicated because of the uNip1j, etc, which is reused internally for Delta.  Maybe try taking out the if statements and calculating everything every time, since usually it all has to be calculated anyhow for Delta.  Or separate the shear calculation into its own routine which can be called from evp directly or from strain_rates_U, similarly to the suggestion above for T.

Reply:
I have refactored strain_rates_T and strain_rates_U.